File: problem/Implement_Router/Router.java
class Router {
    Map<List<Integer>, Integer> mpp = new HashMap<>(); 
    Queue<List<Integer>> queue = new LinkedList<>();
    Map<Integer, List<Integer>> timestamps = new HashMap<>(); 
    Map<Integer, Integer> st = new HashMap<>();
    int maxSize = 0;

    public Router(int memoryLimit) { 
        maxSize = memoryLimit; 
    }

    public boolean addPacket(int source, int destination, int timestamp) {
        List<Integer> packet = Arrays.asList(source, destination, timestamp);

        if (mpp.containsKey(packet))
            return false;
        if (queue.size() == maxSize) { 
            List<Integer> res = queue.poll();
            mpp.remove(res);
            int temp = res.get(1);
            st.put(temp, st.getOrDefault(temp, 0) + 1);
        }
        queue.offer(packet);
        mpp.put(packet, 1);
        timestamps.computeIfAbsent(destination, k -> new ArrayList<>()).add(timestamp);
        return true;
    }

    public int[] forwardPacket() {
        if(queue.isEmpty()) return new int[0];
        List<Integer> res = queue.poll();
        mpp.remove(res);
        int temp = res.get(1);
        st.put(temp, st.getOrDefault(temp, 0) + 1);
        return new int[]{res.get(0), res.get(1), res.get(2)};
    }

    public int getCount(int destination, int startTime, int endTime) {
        if(!timestamps.containsKey(destination))
            return 0;
        List<Integer> p = timestamps.get(destination);
        int temp = st.getOrDefault(destination, 0);
        int right = lowerBound(p, startTime, temp);
        int left = upperBound(p, endTime, temp);
        return left - right;
    }

    private int lowerBound(List<Integer> p, int target, int start) {
        int l = start, r = p.size();
        while(l < r) {
            int mid = (l + r) / 2;
            if(p.get(mid) < target) l = mid + 1;
            else r = mid;
        }
        return l;
    }

    private int upperBound(List<Integer> p, int target, int start) {
        int l = start, r = p.size();
        while(l < r) {
            int mid = (l + r) / 2;
            if(p.get(mid) <= target) l = mid + 1;
            else r = mid;
        }
        return l;
    }
}

File: problem/Implement_Router/Router.py
class Router(object):

    def __init__(self, memoryLimit):
        self.mpp = {} 
        self.queue = deque() 
        self.timestamps = {} 
        self.st = {}
        self.maxSize = memoryLimit 

    def addPacket(self, source, destination, timestamp):
        packet = (source, destination, timestamp)

        if packet in self.mpp:
            return False
        if len(self.queue) == self.maxSize:
            res = self.queue.popleft()
            del self.mpp[res]
            temp = res[1]
            self.st[temp] = self.st.get(temp, 0) + 1
        self.queue.append(packet)
        self.mpp[packet] = 1
        if destination not in self.timestamps:
            self.timestamps[destination] = []
        self.timestamps[destination].append(timestamp)
        return True

    def forwardPacket(self):
        if not self.queue:
            return []
        res = self.queue.popleft()
        del self.mpp[res]
        temp = res[1]
        self.st[temp] = self.st.get(temp, 0) + 1
        return list(res)

    def getCount(self, destination, startTime, endTime):
        if destination not in self.timestamps:
            return 0
        p = self.timestamps[destination]
        temp = self.st.get(destination, 0)
        right = bisect.bisect_left(p, startTime, temp)
        left = bisect.bisect_right(p, endTime, temp)
        return left - right